### PR TITLE
[Graph C API] Enable plugin optimizers and configs.

### PR DESCRIPTION
**Pull request by Solaryee**
This is the 3rd PR(and the last one) of Graph C API, following RFC [Modular TensorFlow Graph C API](https://github.com/tensorflow/community/blob/master/rfcs/20201027-modular-tensorflow-graph-c-api.md).

This PR is mainly to add plugin optimizer and config in meta_optimizer, which includes:
1. Run plugin optimizer in meta_optimizer.
2. Change config if plugin optimizer is enabled, otherwise leave it as default. Print it when plugin config is conflicit with user config.
3. Call `InitGraphPlugin` in `RegisterPluggableDevicePlugin`.

### CODE REVIEW

**Comment by ShengYang1** [comment]
Hi @penpornk @ezhulenev, please help to have a review. Thanks very much!

**Comment by Jianhui-Li** [comment]
@penpornk  This PR is the 3rd Graph C API PR. Without this PR, users won't be able to use graph plugin.

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc [hunk 1]
```diff
@@ -165,10 +165,26 @@ bool MemoryOptimizerEnabled(
   return mem_opt_type != RewriterConfig::NO_MEM_OPT;
 }
 
+Status GetGraphDevice(const GraphDef& g_def, std::set<std::string>& devices) {
```
Comment (ezhulenev) [inline tensorflow/core/grappler/optimizers/meta_optimizer.cc:168]: Output parameters must be taken by pointer.

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc [hunk 2]
```diff
@@ -165,10 +165,26 @@ bool MemoryOptimizerEnabled(
...
+  for (auto node : g_def.node()) {
```
Comment (ezhulenev) [inline tensorflow/core/grappler/optimizers/meta_optimizer.cc:169]: I think `auto` will do a copy here, you'll need `auto&`.

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc [hunk 3]
```diff
@@ -228,72 +253,104 @@ MetaOptimizer::MetaOptimizer(DeviceBase* cpu_device, const ConfigProto& cfg)
...
   }
-  if (!cfg_.disable_model_pruning()) {
+
+  ConfigList plugin_configs = PluginGraphOptimizerRegistry::GetPluginConfigs(
+      cfg_.use_plugin_optimizers() != RewriterConfig::OFF, device_types);
+  if (!cfg_.disable_model_pruning() && !plugin_configs.disable_model_pruning) {
     optimizers->push_back(MakeUnique<ModelPruner>());
   }
-  if (cfg_.implementation_selector() != RewriterConfig::OFF) {
+  if (cfg_.implementation_selector() != RewriterConfig::OFF &&
```
Comment (ezhulenev) [inline tensorflow/core/grappler/optimizers/meta_optimizer.cc:267]: Maybe move this repeated pattern to macro?

```
cfg_.implementation_selector() != RewriterConfig::OFF &&
  plugin_configs.toggle_config["implementation_selector"]
```

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc [hunk 4]
```diff
@@ -376,6 +438,18 @@ Status MetaOptimizer::InitializeCustomGraphOptimizers(
               << optimizer_config.name();
     }
   }
+  return InitializePluginGraphOptimizers(optimizers, device_types);
+}
+
+Status MetaOptimizer::InitializePluginGraphOptimizers(
+    std::vector<std::unique_ptr<GraphOptimizer>>* optimizers,
```
Comment (ezhulenev) [inline tensorflow/core/grappler/optimizers/meta_optimizer.cc:445]: Output parameters should be at the end.

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc [hunk 5]
```diff
@@ -403,6 +477,148 @@ void MetaOptimizer::InitializeVerifiers(
...
+            ? RewriterConfig::ON
+            : RewriterConfig::OFF;
+    user_cfg.toggle_config["auto_parallel"] = cfg_.auto_parallel().enable()
+                                                  ? RewriterConfig::ON
+                                                  : RewriterConfig::OFF;
+    user_cfg.toggle_config["scoped_allocator_optimization"] =
+        cfg_.scoped_allocator_optimization();
+  } else {
+    for (const string& optimizer_name : cfg_.optimizers()) {
+      if (optimizer_name == "pruning") user_cfg.disable_model_pruning = true;
```
Comment (ezhulenev) [inline tensorflow/core/grappler/optimizers/meta_optimizer.cc:534]: Again maybe macro here? Too many duplicated code.

**Review by ezhulenev — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc (hunk 1 above)
Comment (ShengYang1) [reply tensorflow/core/grappler/optimizers/meta_optimizer.cc]: Done

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc (hunk 2 above)
Comment (ShengYang1) [reply tensorflow/core/grappler/optimizers/meta_optimizer.cc]: Done

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc (hunk 3 above)
Comment (ShengYang1) [reply tensorflow/core/grappler/optimizers/meta_optimizer.cc]: The pattern is not exactly the same for different optimizers, so I use 6 macros here.
~~~cpp
#define USER_IS_ON(CFG) cfg_.CFG() == RewriterConfig::ON
#define USER_NOT_OFF(CFG) cfg_.CFG() != RewriterConfig::OFF
#define PLUGIN_IS_ON(CFG) plugin_configs.toggle_config[#CFG] == RewriterConfig::ON
#define PLUGIN_NOT_OFF(CFG) plugin_configs.toggle_config[#CFG] != RewriterConfig::OFF
#define BOTH_IS_ON(CFG) USER_IS_ON(CFG) && PLUGIN_IS_ON(CFG)
#define BOTH_NOT_OFF(CFG) USER_NOT_OFF(CFG) && PLUGIN_NOT_OFF(CFG)
~~~

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc (hunk 4 above)
Comment (ShengYang1) [reply tensorflow/core/grappler/optimizers/meta_optimizer.cc]: OK. Done for all functions.

File: tensorflow/core/grappler/optimizers/meta_optimizer.cc (hunk 5 above)
Comment (ShengYang1) [reply tensorflow/core/grappler/optimizers/meta_optimizer.cc:520]: Done.

**Comment by ShengYang1** [comment]
@ezhulenev Thanks for your review! I have addressed all your comments.

**Review by ezhulenev — APPROVED** [review]

**Comment by gbaned** [comment]
@ShengYang1 Can you please resolve conflicts? Thanks!

**Comment by ShengYang1** [comment]
> @ShengYang1 Can you please resolve conflicts? Thanks!

I have already rebased, please review. Thanks!

**Comment by penpornk** [comment]
#45784 just got [merged](https://github.com/tensorflow/tensorflow/commit/3a3878ff2dba3169c49991552bc1981f53f10099). Could you please resolve conflicts again? Thank you!

**Comment by ShengYang1** [comment]
@penpornk Already rebased. 

Meanwhile I have slightly changed the logic in `RegisterPluggableDevicePlugin`. In the previous design, it will directly return error if `SE_InitPlugin` is not found. However, plugin with graph only should also be allowed, so I changed it to "return error if neither device nor graph is found". Suggestions are welcome. Thanks.

**Review by penpornk — APPROVED** [review]
> Meanwhile I have slightly changed the logic in `RegisterPluggableDevicePlugin`. In the previous design, it will directly return error if `SE_InitPlugin` is not found. However, plugin with graph only should also be allowed, so I changed it to "return error if neither device nor graph is found". Suggestions are welcome. Thanks.

Thank you very much! I was about to ask for that too, because the use case with just the graph plugin is also valid in the original RFC ([2nd scenario](https://github.com/tensorflow/community/blob/master/rfcs/20201027-modular-tensorflow-graph-c-api.md#supported-user-scenarios)).

I marked some trivial typos. No need to fix them. I will fix them internally.

The main PluggableDevice implementation PR (#45784) missed last night's nightly test, so we'll know tomorrow whether it will get reverted. I will merge this PR after I'm sure #45784 won't get reverted.

**Comment by penpornk** [comment]
So I have good news and bad news.
 
**Good news:** This PR is merged, and the [main PluggableDevice PR](https://github.com/tensorflow/tensorflow/pull/45784) is now confirmed safe (based on last night's tests)! 🎉 

**Bad news:** There has been some issues adding the new `RewriterConfig` option (`use_plugin_optimizers`). I added it separately [here](https://github.com/tensorflow/tensorflow/commit/9b6540a862d36acd0b8f294cb2745493acdc13dc) but the change got reverted. So I made this PR use `kUsePluginOptimizers = RewriterConfig::ON;` in `metaoptimizer.cc` and removed the changes in `eager/context.py` for now. I will change them back once I can get the `RewriterConfig` change back in.

**Comment by ShengYang1** [comment]
@penpornk Thanks for your effort to get this PR merged. It's really good news🎉.

**Pull request merged**